src/services/services.ts

```diff
@@ -2912,6 +2919,63 @@ module ts {
 
                 return filteredMembers;
             }
+
+            function isExternalModuleName(token: Node): boolean {
+                if (token && token.parent && token.kind === SyntaxKind.StringLiteral && isInStringOrRegularExpressionOrTemplateLiteral(token)) {
+                    return token.parent.kind === SyntaxKind.ImportDeclaration ||
+                        token.parent.kind === SyntaxKind.ExportDeclaration ||
+                        token.parent.kind === SyntaxKind.ExternalModuleReference;
+                }
+                return false;
+            }
+
+            function getExternalModulecompletions(location: Node): CompletionInfo {
+                var entries: CompletionEntry[] = [];
+                var currentSourceFile = location.getSourceFile();
+                var currentDirectoryPath = getDirectoryPath(normalizePath(currentSourceFile.fileName));
+                var isInAmbientContext = ts.isInAmbientContext(location);
+
+                forEach(program.getSourceFiles(), sourceFile => {
+                    if (isExternalModule(sourceFile)) {
+                        // Only consider these if current request is not in an ambient context,
+                        // and in a diffrent file
+                        if (!isInAmbientContext && sourceFile !== currentSourceFile) {
+                            var relativeModuleName = removeFileExtension(getRelativePathToDirectoryOrUrl(currentDirectoryPath,
+                                sourceFile.fileName, "", getCanonicalFileName, /*isAbsolutePathAnUrl*/ true));
+                            if (relativeModuleName.charCodeAt(0) !== CharacterCodes.dot && relativeModuleName.charCodeAt(1) !== CharacterCodes.slash) {

```

Comment:
I suppose; if it deoptimizes, it may not be as critical as the batch compiler. Still not a painful check though.